### PR TITLE
Preserve single line blocks when emitting.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -2713,7 +2713,21 @@ module ts {
                 emit(node.whenFalse);
             }
 
+            function isSingleLineBlock(node: Node) {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: isSingleLineEmptyBlock

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2886,6 +2900,11 @@ module ts {
                 getLineOfLocalPosition(currentSourceFile, skipTrivia(currentSourceFile.text, node2.pos));
             }
 
+            function nodeEndIsOnSameLineAsNodeStart(node1: Node, node2: Node) {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Does node1 ever differ from node2?

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: If not, then just check the pos and end of the single node

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -3385,73 +3404,79 @@ module ts {
...
+                                decreaseIndent();
+                                emitToken(SyntaxKind.CloseBraceToken, (<Block>node.body).statements.end);
+                            }
+                            else {
+                                decreaseIndent();
+                                emitStart(node.body);
+                                write("}");
+                                emitEnd(node.body);
+                            }
+                        }
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Wow, this diff is awful.

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: I'm assuming nothing changed inside this else block?

**Comment by JsonFreeman** [comment]
:+1:

**Comment by ahejlsberg** [comment]
@CyrusNajmabadi This change has some odd effects on empty functions with default parameter values and/or destructuring patterns. Specifically, the preamble code we generate for those constructs isn't included in an empty function *if the curlies are on the same line*. For example, for this function
```typescript
function foo(x = 0) { }
```
we generate
```typescript
function foo(x) { }
```
but if you put the curlies on different lines we generate
```typescript
function foo(x) {
    if (x === void 0) { x = 0; }
}
```
You could argue that there is no reason to generate the preamble as there are no statements in the function, but it certainly is odd that placing curlies on different lines makes a difference.

**Comment by DanielRosenwasser** [comment]
Actually default initializers can still have side-effects, so if this hasn't been fixed it's a bug.

**Pull request merged**